File: CSharp/src/FAI.Router/JudgeLogic/Specifications.cs
using AI.DataStructs.Algebraic;
using System.Text.Json.Serialization;
using FAI.Router.Enums;

namespace FAI.Router.JudgeLogic;

/// <summary>
/// Спецификация (ТЗ)
/// </summary>
public class Specifications
{
    // Типичный масштаб поля. Вектор сравнивается косинусом, поэтому координаты обязаны быть
    // соизмеримы: в сырых единицах объем в символах давал 98% нормы вектора, и косинус мерил
    // только длину, из-за чего текст в противоположном стиле получал оценку 0,9998.
    private const double SymbolLengthScale = 20000;
    private const double WordLengthScale = 3000;
    private const double CountScale = 20;
    private const double HeadingDepthScale = 6;
    private const double SentenceLengthScale = 40;
    private const double ReadabilityMax = 100;

    /// <summary>
    /// Языки, у которых на арене свой рейтинг, кодами ISO 639-1. Язык из списка светит своим
    /// разрядом, любой другой известный язык светит последним, неизвестный не светит ничем.
    /// </summary>
    public static readonly string[] LanguageCodes = ["en", "ru", "zh", "fr", "de", "es", "ja", "ko", "pl"];

    /// <summary>Разрядов под язык: по одному на язык арены и один на прочие.</summary>
    public static int LanguageDim => LanguageCodes.Length + 1;

    // Доли приходят от модели, а она границы схемы соблюдает не всегда: DeepSeek возвращал
    // termDensity 4 и 80 при объявленных 0-1. Такое значение забивает норму вектора целиком,
    // поэтому границу держит сам тип, а не только схема ответа.
    private double _readabilityScore;
    private double _termDensity;
    private double _formalityScore;
    private double _expertLevel;
    private double _difficulty;
    private double _factualityDemand;

    /// <summary>
    /// Распознаваемый тип стиля (стиль - представлен one-hot вектором)
    /// </summary>
    public Style StyleType { get; set; } = Style.Other;

    /// <summary>
    /// Длинна текста в символах
    /// </summary>
    public int SymbolLength { get; set; }

    /// <summary>
    /// Длинна текста в словах
    /// </summary>
    public int WordLength { get; set; }

    #region Структурные метрики

    /// <summary>
    /// Число абзацев
    /// </summary>
    public int ParagraphCount { get; set; }

    /// <summary>
    /// Число разделов (заголовков верхнего уровня)
    /// </summary>
    public int SectionCount { get; set; }

    /// <summary>
    /// Число пунктов списков
    /// </summary>
    public int ListItemCount { get; set; }

    /// <summary>
    /// Число таблиц
    /// </summary>
    public int TableCount { get; set; }

    /// <summary>
    /// Число блоков кода
    /// </summary>
    public int CodeBlockCount { get; set; }

    /// <summary>
    /// Число формул
    /// </summary>
    public int FormulaCount { get; set; }

    /// <summary>
    /// Глубина вложенности заголовков (H1/H2/H3...)
    /// </summary>
    public int HeadingDepth { get; set; }

    #endregion

    #region Лексико-стилевые метрики

    /// <summary>
    /// Средняя длина предложения (в словах)
    /// </summary>
    public double AvgSentenceLength { get; set; }

    /// <summary>
    /// Читаемость текста (индекс Флеша-Кинкейда или аналог), 0-100
    /// </summary>
    public double ReadabilityScore
    {
        get => _readabilityScore;
        set => _readabilityScore = Math.Clamp(value, 0, ReadabilityMax);
    }

    /// <summary>
    /// Доля терминологии/иностранных слов, 0-1
    /// </summary>
    public double TermDensity
    {
        get => _termDensity;
        set => _termDensity = Math.Clamp(value, 0, 1);
    }

    /// <summary>
    /// Формальность тона, 0-1
    /// </summary>
    public double FormalityScore
    {
        get => _formalityScore;
        set => _formalityScore = Math.Clamp(value, 0, 1);
    }

    #endregion

    #region Соответствие заказу

    /// <summary>
    /// Язык ответа
    /// </summary>
    public string? Language { get; set; }

    /// <summary>
    /// Наличие ссылок/источников
    /// </summary>
    public bool HasReferences { get; set; }

    #endregion

    #region Предмет задачи

    /// <summary>
    /// Предметная область
    /// </summary>
    public Domain Domain { get; set; } = Domain.General;

    /// <summary>
    /// Язык программирования, если заказан или написан код
    /// </summary>
    public ProgrammingLanguage ProgrammingLanguage { get; set; } = ProgrammingLanguage.None;

    /// <summary>
    /// Область науки, если задача научная
    /// </summary>
    public ScienceField ScienceField { get; set; } = ScienceField.None;

    /// <summary>
    /// Тип задачи: что заказчик хочет получить на выходе (письмо, отчет, лендинг, код)
    /// </summary>
    public TaskKind TaskKind { get; set; } = TaskKind.None;

    #endregion

    #region Требования заказа вне вектора ответа

    /// <summary>
    /// Насколько запрос требует экспертной подготовки, 0-1 (категория арены Expert)
    /// </summary>
    public double ExpertLevel
    {
        get => _expertLevel;
        set => _expertLevel = Math.Clamp(value, 0, 1);
    }

    /// <summary>
    /// Трудность запроса: доля из семи признаков трудного запроса арены, 0-1 (Hard Prompts)
    /// </summary>
    public double Difficulty
    {
        get => _difficulty;
        set => _difficulty = Math.Clamp(value, 0, 1);
    }

    /// <summary>
    /// Насколько ответ держится на проверяемых фактах, 0-1: чем выше, тем дороже ошибка в факте
    /// </summary>
    public double FactualityDemand
    {
        get => _factualityDemand;
        set => _factualityDemand = Math.Clamp(value, 0, 1);
    }

    /// <summary>
    /// Смысловые пункты, которые ответ обязан раскрыть по сути: что сравнить, посчитать, решить.
    /// По ним судья содержания проверяет полноту, а не число разделов.
    /// </summary>
    public List<string> RequiredPoints { get; set; } = [];

    /// <summary>
    /// Явные ограничения запроса, выполнение которых можно проверить (Instruction Following)
    /// </summary>
    public List<string> Constraints { get; set; } = [];

    #endregion

    /// <summary>
    /// Вектор признаков
    /// </summary>
    /// <remarks>
    /// Вычисляется из остальных полей, поэтому в сохраненный вид не попадает: иначе накопитель
    /// хранил бы одни и те же данные дважды, а при смене масштабов старые копии разошлись бы
    /// с тем, что дает текущий код.
    /// </remarks>
    [JsonIgnore]
    public Vector FeaturesSpecificationVector => GetVector();

    /// <summary>
    /// Разряд языка в векторе: по списку <see cref="LanguageCodes"/>, последний для прочих,
    /// минус единица для неизвестного
    /// </summary>
    /// <param name="code">Код языка ISO 639-1</param>
    public static int LanguageSlot(string? code)
    {
        if (string.IsNullOrWhiteSpace(code))
            return -1;

        int index = Array.IndexOf(LanguageCodes, code.Trim().ToLowerInvariant());

        return index >= 0 ? index : LanguageCodes.Length;
    }

    // Формирования вектора признаков
    private Vector GetVector()
    {
        // Предмет задачи, язык и ссылки идут кодом «один из многих» и признаком 0/1: по ним
        // роутер учит, кто в какой области силен, а масштабов у них нет, как и у стиля
        Vector featuresVector =
        [
            .. OneHot(StyleType),
            Scaled(SymbolLength, SymbolLengthScale),
            Scaled(WordLength, WordLengthScale),
            Scaled(ParagraphCount, CountScale),
            Scaled(SectionCount, CountScale),
            Scaled(ListItemCount, CountScale),
            Scaled(TableCount, CountScale),
            Scaled(CodeBlockCount, CountScale),
            Scaled(FormulaCount, CountScale),
            Scaled(HeadingDepth, HeadingDepthScale),
            Scaled(AvgSentenceLength, SentenceLengthScale),
            ReadabilityScore / ReadabilityMax,
            TermDensity,
            FormalityScore,
            .. Subject(Domain),
            .. Subject(ProgrammingLanguage),
            .. Subject(ScienceField),
            .. Subject(TaskKind),
            .. LanguageVector(),
            HasReferences ? 1 : 0
        ];
        return featuresVector;
    }

    // Логарифмическая шкала: у объемов и счетчиков значимо отношение величин, а не разница,
    // а деление на масштаб приводит поле к единичному порядку. Отрицательное значение может
    // прийти от модели, распознающей ТЗ, а логарифм на нем дает NaN и портит весь вектор.
    internal static double Scaled(double value, double scale) =>
        Math.Log(1 + Math.Max(0, value)) / Math.Log(1 + scale);

    private Vector LanguageVector()
    {
        Vector vector = new(LanguageDim);
        int slot = LanguageSlot(Language);

        if (slot >= 0)
            vector[slot] = 1;

        return vector;
    }

    // Код «один из многих» по перечислению: разряд по порядку значения
    private static Vector OneHot<TEnum>(TEnum value) where TEnum : struct, Enum
    {
        Vector vector = new(Enum.GetValues<TEnum>().Length);
        vector[Convert.ToInt32(value)] = 1;
        return vector;
    }

    // Код предмета задачи: как «один из многих», но первое значение означает «не задано» и
    // разряда не имеет. Задача без предмета получает те же координаты, что и раньше, и оценки
    // судьи на ней не меняются
    private static Vector Subject<TEnum>(TEnum value) where TEnum : struct, Enum
    {
        Vector vector = new(Enum.GetValues<TEnum>().Length - 1);
        int index = Convert.ToInt32(value) - 1;

        if (index >= 0)
            vector[index] = 1;

        return vector;
    }
}

File: Python/fai_router/specifications.py
from __future__ import annotations

import math
from typing import Any

import numpy as np

from fai_router.enums import Domain, Indexed, ProgrammingLanguage, ScienceField, Style, TaskKind


class Specifications:
    """Спецификация (техническое задание). Один и тот же тип описывает и требования к ответу,
    и свойства готового текста, поэтому заказ и факт можно сравнить."""

    # Типичный масштаб поля. Вектор сравнивается косинусом, поэтому координаты обязаны быть
    # соизмеримы: в сырых единицах объем в символах давал 98% длины вектора, и косинус мерил
    # только длину, из-за чего текст в противоположном стиле получал оценку 0,9998
    SYMBOL_LENGTH_SCALE = 20000.0
    WORD_LENGTH_SCALE = 3000.0
    COUNT_SCALE = 20.0
    HEADING_DEPTH_SCALE = 6.0
    SENTENCE_LENGTH_SCALE = 40.0
    READABILITY_MAX = 100.0

    # Языки, у которых на арене свой рейтинг, кодами ISO 639-1. Язык из списка светит своим
    # разрядом, любой другой известный язык светит последним, неизвестный не светит ничем
    LANGUAGE_CODES = ("en", "ru", "zh", "fr", "de", "es", "ja", "ko", "pl")

    def __init__(
        self,
        style_type: Style = Style.OTHER,
        symbol_length: int = 0,
        word_length: int = 0,
        paragraph_count: int = 0,
        section_count: int = 0,
        list_item_count: int = 0,
        table_count: int = 0,
        code_block_count: int = 0,
        formula_count: int = 0,
        heading_depth: int = 0,
        avg_sentence_length: float = 0.0,
        readability_score: float = 0.0,
        term_density: float = 0.0,
        formality_score: float = 0.0,
        language: str | None = None,
        has_references: bool = False,
        domain: Domain = Domain.GENERAL,
        programming_language: ProgrammingLanguage = ProgrammingLanguage.NONE,
        science_field: ScienceField = ScienceField.NONE,
        task_kind: TaskKind = TaskKind.NONE,
        expert_level: float = 0.0,
        difficulty: float = 0.0,
        factuality_demand: float = 0.0,
        required_points: list[str] | None = None,
        constraints: list[str] | None = None,
    ):
        self.style_type = style_type
        self.symbol_length = symbol_length
        self.word_length = word_length
        self.paragraph_count = paragraph_count
        self.section_count = section_count
        self.list_item_count = list_item_count
        self.table_count = table_count
        self.code_block_count = code_block_count
        self.formula_count = formula_count
        self.heading_depth = heading_depth
        self.avg_sentence_length = avg_sentence_length
        self.readability_score = readability_score
        self.term_density = term_density
        self.formality_score = formality_score
        self.language = language
        self.has_references = has_references
        # Предмет задачи: по нему роутер учит, кто в какой области силен
        self.domain = domain
        self.programming_language = programming_language
        self.science_field = science_field
        self.task_kind = task_kind
        # Требования заказа вне вектора ответа: у готового текста их нет, поэтому в вектор
        # спецификации они не входят, а читает их InputFeatures как свойства задачи
        self.expert_level = expert_level
        self.difficulty = difficulty
        self.factuality_demand = factuality_demand
        # Смысловые пункты и явные ограничения: по ним судья содержания проверяет полноту и
        # выполнение указаний, а не число разделов
        self.required_points = list(required_points or [])
        self.constraints = list(constraints or [])

    # Доли приходят от модели, а она границы схемы соблюдает не всегда: DeepSeek возвращал
    # term_density 4 и 80 при объявленных 0..1. Такое значение забивает длину вектора целиком,
    # поэтому границу держит сам тип, а не только схема ответа.

    @property
    def readability_score(self) -> float:
        return self._readability

    @readability_score.setter
    def readability_score(self, value: float) -> None:
        self._readability = min(max(float(value), 0.0), self.READABILITY_MAX)

    @property
    def term_density(self) -> float:
        return self._term_density

    @term_density.setter
    def term_density(self, value: float) -> None:
        self._term_density = min(max(float(value), 0.0), 1.0)

    @property
    def formality_score(self) -> float:
        return self._formality

    @formality_score.setter
    def formality_score(self, value: float) -> None:
        self._formality = min(max(float(value), 0.0), 1.0)

    @property
    def expert_level(self) -> float:
        """Насколько запрос требует экспертной подготовки, 0..1 (категория арены Expert)."""
        return self._expert_level

    @expert_level.setter
    def expert_level(self, value: float) -> None:
        self._expert_level = min(max(float(value), 0.0), 1.0)

    @property
    def difficulty(self) -> float:
        """Доля из семи признаков трудного запроса арены, 0..1 (Hard Prompts)."""
        return self._difficulty

    @difficulty.setter
    def difficulty(self, value: float) -> None:
        self._difficulty = min(max(float(value), 0.0), 1.0)

    @property
    def factuality_demand(self) -> float:
        """Насколько ответ держится на проверяемых фактах, 0..1."""
        return self._factuality_demand

    @factuality_demand.setter
    def factuality_demand(self, value: float) -> None:
        self._factuality_demand = min(max(float(value), 0.0), 1.0)

    @classmethod
    def language_dim(cls) -> int:
        """Разрядов под язык: по одному на язык арены и один на прочие."""
        return len(cls.LANGUAGE_CODES) + 1

    @classmethod
    def language_slot(cls, code: str | None) -> int:
        """Разряд языка в векторе: по списку LANGUAGE_CODES, последний для прочих, минус единица
        для неизвестного."""
        if not code or not code.strip():
            return -1
        code = code.strip().lower()
        return cls.LANGUAGE_CODES.index(code) if code in cls.LANGUAGE_CODES else len(cls.LANGUAGE_CODES)

    def feature_vector(self) -> np.ndarray:
        """Вектор признаков: коды «один из многих» стиля и предмета задачи, приведенные к масштабу
        метрики, язык и признак ссылок. У предмета, языка и ссылок масштабов нет, как и у стиля."""
        scaled = self.scaled
        numeric = np.array([
            scaled(self.symbol_length, self.SYMBOL_LENGTH_SCALE),
            scaled(self.word_length, self.WORD_LENGTH_SCALE),
            scaled(self.paragraph_count, self.COUNT_SCALE),
            scaled(self.section_count, self.COUNT_SCALE),
            scaled(self.list_item_count, self.COUNT_SCALE),
            scaled(self.table_count, self.COUNT_SCALE),
            scaled(self.code_block_count, self.COUNT_SCALE),
            scaled(self.formula_count, self.COUNT_SCALE),
            scaled(self.heading_depth, self.HEADING_DEPTH_SCALE),
            scaled(self.avg_sentence_length, self.SENTENCE_LENGTH_SCALE),
            self.readability_score / self.READABILITY_MAX,
            self.term_density,
            self.formality_score,
        ])
        language = np.zeros(self.language_dim())
        slot = self.language_slot(self.language)
        if slot >= 0:
            language[slot] = 1.0
        return np.concatenate([
            _one_hot(self.style_type), numeric,
            _subject(self.domain), _subject(self.programming_language),
            _subject(self.science_field), _subject(self.task_kind),
            language, [1.0 if self.has_references else 0.0],
        ])

    @staticmethod
    def scaled(value: float, scale: float) -> float:
        """Логарифмическая шкала: у объемов и счетчиков значимо отношение величин, а не
        разница, а деление на масштаб приводит поле к единичному порядку. Отрицательное
        значение может прийти от модели, а логарифм на нем дает NaN и портит весь вектор."""
        return math.log(1 + max(0.0, float(value))) / math.log(1 + scale)

    _FIELDS = (
        "symbol_length", "word_length", "paragraph_count", "section_count",
        "list_item_count", "table_count", "code_block_count", "formula_count",
        "heading_depth", "avg_sentence_length", "readability_score", "term_density",
        "formality_score", "language", "has_references",
        "expert_level", "difficulty", "factuality_demand",
    )

    _LIST_FIELDS = ("required_points", "constraints")

    # Поля-перечисления пишутся значениями, как в версии на C#; старые записи без них читаются
    # как «не задано»
    _ENUM_FIELDS = {
        "style_type": Style, "domain": Domain, "programming_language": ProgrammingLanguage,
        "science_field": ScienceField, "task_kind": TaskKind,
    }

    def to_dict(self) -> dict[str, Any]:
        data = {name: getattr(self, name) for name in self._FIELDS}
        for name in self._ENUM_FIELDS:
            data[name] = getattr(self, name).value
        for name in self._LIST_FIELDS:
            data[name] = list(getattr(self, name))
        return data

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "Specifications":
        spec = cls()
        for name in cls._FIELDS:
            if name in data and data[name] is not None:
                setattr(spec, name, data[name])
        for name, enum_cls in cls._ENUM_FIELDS.items():
            value = data.get(name)
            if value is not None:
                setattr(spec, name, enum_cls(value))
        for name in cls._LIST_FIELDS:
            setattr(spec, name, [str(item) for item in data.get(name) or []])
        return spec


def _one_hot(value: Indexed) -> np.ndarray:
    """Код «один из многих» по перечислению: разряд по порядку значения."""
    vector = np.zeros(len(type(value)))
    vector[value.index] = 1.0
    return vector


def _subject(value: Indexed) -> np.ndarray:
    """Код предмета задачи: как «один из многих», но первое значение означает «не задано» и
    разряда не имеет. Задача без предмета получает те же координаты, что и раньше, и оценки
    судьи на ней не меняются."""
    vector = np.zeros(len(type(value)) - 1)
    if value.index > 0:
        vector[value.index - 1] = 1.0
    return vector
